actors/market/src/lib.rs

```diff
@@ -275,64 +275,56 @@ impl Actor {
+            // Fetch client's datacap balance and calculate the amount of datacap required for the verified deals.
+            // Drop any verified deals for which the client has insufficient datacap.
+            if deal.proposal.verified_deal {
+                let mut remaining_datacap =
+                    match all_client_remaining_datacap.get(&client_id).cloned() {
+                        None => {
+                            balance_of(rt, &Address::new_id(client_id)).map_err(
+                                |e| actor_error!(not_found; "failed to get datacap {}", e.msg()),
+                            )?
                         }
-                        ids[0]
-                    }
-                    Err(e) => {
-                        info!(
-                            "invalid deal {}: failed to allocate datacap for verified deal: {}",
-                            di, e
-                        );
-                        continue;
-                    }
+                        Some(client_data) => client_data,
+                    };
+                let piece_datacap_required =
+                    TokenAmount::from_whole(deal.proposal.piece_size.0 as i64);
+                if remaining_datacap < piece_datacap_required {
+                    continue;
                 }
-            } else {
-                NO_ALLOCATION_ID
-            };
+                remaining_datacap -= &piece_datacap_required;
+                all_client_remaining_datacap.insert(client_id, remaining_datacap);
+                client_alloc_reqs.entry(client_id).or_default().push(alloc_request_for_deal(
+                    &deal,
+                    rt.policy(),
+                    curr_epoch,
+                ));
+            }
 
             total_provider_lockup = provider_lockup;
             total_client_lockup.insert(client_id, client_lockup);
             proposal_cid_lookup.insert(pcid);
-            valid_deals.push(ValidDeal {
-                proposal: deal.proposal,
-                cid: pcid,
-                allocation: allocation_id,
-            });
+            valid_deals.push(ValidDeal { proposal: deal.proposal, cid: pcid });
             valid_input_bf.set(di as u64)
         }
 
+        let mut client_allocations: BTreeMap<ActorID, Vec<AllocationID>> = BTreeMap::new();
+        for (client_id, alloc_reqs) in client_alloc_reqs.iter() {
+            let params =
+                datacap_transfer_request(&Address::new_id(*client_id), alloc_reqs.clone())?;
+            let mut alloc_ids = match transfer_from(rt, params) {
+                Ok(ids) => ids,
+                Err(e) => {
+                    return Err(actor_error!(
+                        illegal_state,
+                        "failed transfer datacap: {}",
+                        e.msg()
+                    ));
+                }
+            };
+            client_allocations.entry(*client_id).or_default().append(alloc_ids.as_mut());
```

Comment:
This should be just a simple `insert`. There is no possibility of an existing entry.